### PR TITLE
network: compare addresses more strictly

### PR DESCRIPTION
**Pull request by yuwata**
This re-introduce the full hash and compre functions for Address,
which was reverted https://github.com/systemd/systemd/commit/1d30fc5cb64ecba2f03fe42aa0d8c65c3decad82 (https://github.com/systemd/systemd/pull/17851).

The issue https://github.com/systemd/systemd/issues/17831, which is fixed by https://github.com/systemd/systemd/pull/17851, is handled in a different way;
simply ignore to configure the conflicted DHCPv6 address. Previously, we
warn about the conflict, but continue to configure the address anyway, and
the kernel ignores the request. So, it is not necessary to request the
conflicted address in networkd side.

This fixes the following issues:
- when a link has an address, and corresponding .network file has the
  address with different prefix length, then the prefix length specified
  in the .network file will not be applied,
- we cannot specify multiple IPv4 addresses with different prefix
  length, e.g.
  ```
  Address=10.10.10.10/24
  Address=10.10.10.10/16
  ```
  This is spurious setup, but the kernel allows it.

Fixes https://github.com/systemd/systemd/issues/22515.

### CODE REVIEW

**Comment by bluca** [comment]
```
 391/1264 test-network                                                                                                                    FAIL             0.27s   killed by signal 6 SIGABRT
>>> MALLOC_PERTURB_=85 PATH=/home/runner/work/systemd/systemd/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin SYSTEMD_LANGUAGE_FALLBACK_MAP=/home/runner/work/systemd/systemd/src/locale/language-fallback-map /home/runner/work/systemd/systemd/build/test-network
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
Assertion 'address_equal(a1, a2)' failed at src/network/test-network.c:211, function test_address_equality(). Aborting.
```

**Comment by keszybz** [comment]
`subprocess.CalledProcessError: Command '['/systemd-meson-build/systemd-networkd-wait-online', '--timeout=20s', '--interface=wg99:routable', '--interface=wg98:routable', '--interface=wg97:carrier']' returned non-zero exit status 1.`
Hmm.

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -1980,8 +2011,10 @@ int network_drop_invalid_addresses(Network *network) {
                         address_free(dup);
                 }
 
-                /* Do not use address_hash_ops_free here. Otherwise, all address settings will be freed. */
-                r = set_ensure_put(&addresses, &address_hash_ops, address);
+                /* Use address_kernel_hash_ops here. The function address_kernel_compare_func() matches
+                 * how kernel compares addresses, and is more lenient than address_compare_func().
+                 * Hence, the logic of dedup here is stricter than the case address_hash_ops is used. */
```
Comment (keszybz) [inline src/network/networkd-address.c]: … than when address_hash_ops is used.

File: src/network/networkd-address.c [hunk 2]
```diff
@@ -496,92 +550,69 @@ int address_get(Link *link, const Address *in, Address **ret) {
...
+                        continue;
+
+                if (in_addr_is_set(family, &a->in_addr_peer))
+                        continue;
+
+                if (ret)
+                        *ret = a;
+
+                return 0;
         }
```
Comment (keszybz) [inline src/network/networkd-address.c:600]: Maybe add a comment explaining the logic here? It's not obvious why `in_addr_peer` is checked.

File: src/network/networkd-dhcp6.c [hunk 3]
```diff
@@ -181,20 +182,21 @@ static void log_dhcp6_address(Link *link, const Address *address) {
         if (existing->source == NETWORK_CONFIG_SOURCE_NDISC)
                 by_ndisc = true;
 
-        log_link_warning(link, "DHCPv6 address %s/%u (valid %s, preferred %s) conflicts the address %s/%u%s.",
+        log_link_warning(link, "DHCPv6 address %s/%u (valid %s, preferred %s) conflicts the address %s/%u%s. Ignoring the DHCPv6 address.",
```
Comment (keszybz) [inline src/network/networkd-dhcp6.c]: Pre-existing condition, but "a conflicts b" is very awkward. "a conflicts with b" would be better.
Maybe change this to:
```C
log_link_warning(link, 
                        "Ignoring DHCPv6 address %s/%u (valid %s, preferred %s) which conflicts with %s/%u%s.",
                         strna(buffer), address->prefixlen,
                         FORMAT_LIFETIME(address->lifetime_valid_usec),
                         FORMAT_LIFETIME(address->lifetime_preferred_usec),
                         strna(buffer), existing->prefixlen,
                         by_ndisc ? " assigned by NDisc" : "");
if (by_ndisc)
   log_link_info(link, "Hint: use IPv6Token= setting to change the address generated by NDISC or set UseAutonomousPrefix=no.");
```

**Review by keszybz — COMMENTED** [review]
Looks reasonable, but I'm not familiar with the details here. The CI failure looks like it could be related.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. Addressed the above comments.
Now, cleanups for broadcast handling is added, so the failure in the test for wireguard should be fixed.
PTAL.

**Comment by keszybz** [comment]
focal-ppc64el: TEST-50-DISSECT:                    FAIL     (610 s)
focal-s390x: something in unittests, but I don't see what
rpm-build:https://bugzilla.redhat.com/show_bug.cgi?id=2057735
centos ci: TEST-58-REPART

This looks unrelated.

**Comment by keszybz** [comment]
LGTM.

**Pull request merged**